esp32s3/src/usb0.rs

```diff
@@ -686,396 +588,128 @@ impl RegisterBlock {
-    #[inline(always)]
-    pub const fn dtxfsts5(&self) -> &DTXFSTS5 {
-        &self.dtxfsts5
-    }
-    #[doc = "0x9bc - "]
-    #[inline(always)]
-    pub const fn diepdmab5(&self) -> &DIEPDMAB5 {
-        &self.diepdmab5
-    }
-    #[doc = "0x9c0 - "]
-    #[inline(always)]
-    pub const fn diepctl6(&self) -> &DIEPCTL6 {
-        &self.diepctl6
-    }
-    #[doc = "0x9c8 - "]
-    #[inline(always)]
-    pub const fn diepint6(&self) -> &DIEPINT6 {
-        &self.diepint6
-    }
-    #[doc = "0x9d0 - "]
-    #[inline(always)]
-    pub const fn dieptsiz6(&self) -> &DIEPTSIZ6 {
-        &self.dieptsiz6
-    }
-    #[doc = "0x9d4 - "]
-    #[inline(always)]
-    pub const fn diepdma6(&self) -> &DIEPDMA6 {
-        &self.diepdma6
-    }
-    #[doc = "0x9d8 - "]
-    #[inline(always)]
-    pub const fn dtxfsts6(&self) -> &DTXFSTS6 {
-        &self.dtxfsts6
-    }
-    #[doc = "0x9dc - "]
-    #[inline(always)]
-    pub const fn diepdmab6(&self) -> &DIEPDMAB6 {
-        &self.diepdmab6
-    }
-    #[doc = "0xb00 - "]
-    #[inline(always)]
-    pub const fn doepctl0(&self) -> &DOEPCTL0 {
-        &self.doepctl0
-    }
-    #[doc = "0xb08 - "]
-    #[inline(always)]
-    pub const fn doepint0(&self) -> &DOEPINT0 {
-        &self.doepint0
-    }
-    #[doc = "0xb10 - "]
-    #[inline(always)]
-    pub const fn doeptsiz0(&self) -> &DOEPTSIZ0 {
-        &self.doeptsiz0
-    }
-    #[doc = "0xb14 - "]
-    #[inline(always)]
-    pub const fn doepdma0(&self) -> &DOEPDMA0 {
-        &self.doepdma0
-    }
-    #[doc = "0xb1c - "]
-    #[inline(always)]
-    pub const fn doepdmab0(&self) -> &DOEPDMAB0 {
-        &self.doepdmab0
-    }
-    #[doc = "0xb20 - "]
-    #[inline(always)]
-    pub const fn doepctl1(&self) -> &DOEPCTL1 {
-        &self.doepctl1
-    }
-    #[doc = "0xb28 - "]
-    #[inline(always)]
-    pub const fn doepint1(&self) -> &DOEPINT1 {
-        &self.doepint1
-    }
-    #[doc = "0xb30 - "]
+    #[doc = "0x900..0x920 - Device IN endpoint 0"]
     #[inline(always)]
-    pub const fn doeptsiz1(&self) -> &DOEPTSIZ1 {
-        &self.doeptsiz1
+    pub const fn in_ep0(&self) -> &IN_EP0 {
+        &self.in_ep0
     }
-    #[doc = "0xb34 - "]
+    #[doc = "0x920..0x9e0 - Device IN endpoints 1-6"]
     #[inline(always)]
-    pub const fn doepdma1(&self) -> &DOEPDMA1 {
-        &self.doepdma1
+    pub const fn in_ep(&self, n: usize) -> &IN_EP {
+        &self.in_ep[n]
     }
-    #[doc = "0xb3c - "]
+    #[doc = "Iterator for array of:"]
+    #[doc = "0x920..0x9e0 - Device IN endpoints 1-6"]
     #[inline(always)]
-    pub const fn doepdmab1(&self) -> &DOEPDMAB1 {
-        &self.doepdmab1
+    pub fn in_ep_iter(&self) -> impl Iterator<Item = &IN_EP> {
+        self.in_ep.iter()
     }
-    #[doc = "0xb40 - "]
+    #[doc = "0x920..0x940 - Device IN endpoints 1-6"]
     #[inline(always)]
-    pub const fn doepctl2(&self) -> &DOEPCTL2 {
-        &self.doepctl2
+    pub const fn in_ep1(&self) -> &IN_EP {
+        self.in_ep(0)
```

Comment:
The endpoint clustering does have some usability issues that I'm not thrilled about.  Because Endpoint 0 has slightly different register fields than the other endpoints, it cannot be clustered together with them.  svd2rust produces a somewhat unfortunate API as a result:

The `in_ep0()`, `in_ep1()` ... `in_ep6()` methods do what you expect.
However, the `in_ep(n: usize)` method is slightly confusing since it returns the endpoint 1 less than the argument : calling `in_ep(2)` returns registers for endpoint 1.

The same applies to the `out_ep()` method.

Reply:
Hmm that is quite unfortunate. 

If I understand correctly, ep0 is still a 32 bit register, it's just that some of the fields are missing or have a smaller bit width. I wonder if we could just treat ep0 as any other ep and being careful to only write the correct bits for ep0 specifically in the impl. Expanding on that, we could keep the register specification for ep0 around (if that's possible with svd2rust), and cast the register we get from `in_ep(usize)`, something like `&IN_EP.as_ptr() as *const IN_EP0`. This still isn't ideal, because the impl can forget to do the cast and end up writing the wrong bits, but it might be better? What do you think?

Reply:
I do think it is useful to have separate register definitions for endpoint 0: the behavior is meaningfully different in some notable cases.  The fact that Rust has different register types makes it easier to avoid making mistakes here.

For one particular example: in DIEPCTLx the MPS field is 11 bits wide, and you write the endpoint's maximum packet size here.  However, for DIEPCTL0 the MPS field is only 2 bits wide, and you write the MPS effectively as an enum (0 means 64 bytes, 1 means 32 bytes, 2 means 16 bytes, and 3 means 8 bytes).  If you don't pay attention to this an try to write the numerical MPS value to DIEPCTL0 you might get surprising results.  (That said, in almost all cases you are going to want EP0's max packet size as 64, and you would luck out with that value since the lower 2 bits are zeros.)

In this most recent update I have used `deriveFrom` to make all endpoints share the same register type definitions when the register definitions are the same, and only use different types when EP0 has different definitions.

The indexing behavior of `in_ep()` is rather unfortunate.  That said, I'm sort of inclined to just live with it for now.  We could provide our own alternative API with some sort of Usb0Ext trait, but that doesn't really help hide this existing confusing API.  In my code for now I have simply been calling `in_ep(ep_index - 1)`, and I don't think that has been too horrible.

svd2rust does this in other places: anywhere that you have a cluster with dimIndex starting at something other than 0. e.g., func_in_sel_cfg_iter() in esp32p4/src/gpio.rs appears to behave this way.